docs/latest/design/envoy-patch-policy.md

```diff
@@ -0,0 +1,150 @@
+# EnvoyPatchPolicy
+
+## Overview
+
+This design introduces the `EnvoyPatchPolicy` API allowing users to modify the generated Envoy xDS Configuration
+that Envoy Gateway generates before sending it to Envoy Proxy.
+
+Envoy Gateway allows users to configure networking and security intent using the
+upstream [Gateway API][] as well as implementation specific [Extension APIs][] defined in this project
+to provide a more batteries included experience for application developers.
+* These APIs are an abstracted version of the underlying Envoy xDS API to provide a a better user experience for the application developer, exposing and setting only a subset of the fields for a specific feature, sometimes in a opinionated way (e.g [RateLimit][])
+* These APIs do not expose all the features capabilities that Envoy has either because these features are desired but the API
+is not defined yet or the project cannot support such an extensive list of features.
+To alleviate this problem, and provide an interim solution for a small section of advanced users who are well versed in
+Envoy xDS API and its capabilities, this API is being introduced.
+
+## Goals
+* Add an API allowing users to modify the generated xDS Configuration
+
+## Non Goals
+* Support multiple patch mechanisims
+
+## Implementation
+`EnvoyPatchPolicy` is a [Direct Policy Attachment][] type API that can be used to extend [Gateway API][]
+Modifications to the generated xDS configuration can be provided as a JSON Patch which is defined in
+[RFC 6902][]. This patching mechanism has been adopted in [Kubernetes][] as well as [Kustomize][] to update
+resource objects.
+
+### Example
+Here is an example highlighting how a user can configure global ratelimiting using an external rate limit service using this API.
+
+```
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: eg-gw
+  namespace: default
+spec:
+  gatewayClassName: eg-gc
+  listeners:
+    - name: example
+      protocol: HTTPS
+      port: 443
+      hostname: example.com
+      tls:
+        certificateRefs:
+          - kind: Secret
+            group: ""
+            name: example-cert
+---            
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: EnvoyPatchPolicy
+metadata:
+  name: ratelimit-patch-policy
+  namespace: default
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io/v1beta1
+    kind: Gateway
+    name: eg-gw
+    namespace: default
+  type: JSONPatch
+  jsonPatches:
+    - type: "type.googleapis.com/envoy.config.listener.v3.Listener"
+      # The listener name is of the form <GatewayNamespace>-<GatewayName>-<GatewayListenerName>
+      name: default-eg-gw-example
+      operation:
+        op: add
+        path: "/filter_chains/0/filters/0/http_filters/0"
+        value: |
+          name: "envoy.filters.http.ratelimit"
+          typed_config:
+            "@type": "type.googleapis.com/envoy.extensions.filters.http.ratelimit.v3.RateLimit"
```

Comment:
yeah we autogenerate and import today in the client side (`egctl`)
with this PR we will need to do it in server side too
https://github.com/envoyproxy/gateway/blob/main/internal/xds/extensions/extensions.gen.go

Reply:
:+1: one minor suggestion: don't import the v2 stuff. One less headache to deal with down the line... and in only sort of works IME

Reply:
thanks,  will rm them

Reply:
I guess we can still use customized extensions, right? It's really important because most EG deployments would have some sort of customized extensions which don't exist in the Envoy upstream.

I have done similar things in Istio with “typed_config". Below is an example:

```json
{
  "name":"envoy.filters.network.meta_protocol_proxy",
  "typed_config":{
    "@type":"type.googleapis.com/udpa.type.v1.TypedStruct",
    "type_url":"type.googleapis.com/aeraki.meta_protocol_proxy.v1alpha.MetaProtocolProxy",
     ....
  }
}

Since it's wrapped in a typed_config, Even Istio doesn't know about the MetaProtocolProxy API, it won't complain about it.
```

Reply:
raised https://github.com/envoyproxy/gateway/pull/1442 to rm v2 imports

Reply:
@zhaohuabing I dont think we'll be able to handle your use case if we purely relied on `protojson` tooling

Reply:
It would be a headache for any user who adds their own custom filters to envoy since EG can't recognize the API of these filters.

Reply:
I understand your sentiment, and it is something we should support, which can be tracked with a GH issue, for now, it doesnt require any changes to the API, or am I missing something ?

Reply:
Yes, it's an implementation concern rather than API.